Use AOTI as inductor backend with precompile mode.

zhxchen17 opened the pull request:
Summary:
Design doc: https://docs.google.com/document/d/1Z15cBBPjoZ7gH00TSgCdgaYko7a7Br-ERd3_hA-g2IU/edit?usp=sharing

In this diff we are trying to introduce some stateful API to enable a global mode which will force inductor to use AOTI as a backend. Different from PR https://github.com/pytorch/pytorch/pull/141700, we didn't try to populate the package file into caching system, instead we bypass caching to simplify the implementation in the current form.

Similar to PR https://github.com/pytorch/pytorch/pull/141700, I did a quick benchmark to the loading time and it looks like the following:
- Precompile
```
buck run mode/opt scripts/zhxchen17:precompile
```
- Load using cache:
```
time buck run mode/opt scripts/zhxchen17:precompile -- --loader cache
```
Output:
```
real    0m24.593s
user    0m59.342s
sys     0m17.201s
```
- Load using load_fullgraph_package
```
time buck run mode/opt scripts/zhxchen17:precompile -- --loader precompile
```
Output:
```
real    0m10.907s
user    0m9.210s
sys     0m1.173s
```

Test Plan:
buck run mode/opt caffe2/test:test_export -- -r test_fullgraph_package_basic
_function

Differential Revision: D68459341


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

ezyang commented on torch/_fullgraph.py:
```diff
@@ -0,0 +1,52 @@
+import dataclasses
```
Why putting this at the very top level torch namespace?

zhxchen17 replied:
No particular reason. I can move it to another place.

jamesjwu commented on torch/__init__.py:
```diff
@@ -2275,11 +2275,12 @@ def compiled_with_cxx11_abi() -> builtins.bool:
 
         self.config: dict[str, _Any] = {}
         self.dynamic = dynamic
+        self.name = name
```
I expect that changing the torch.compile() API will require more people to look at this than just me haha, cc @ezyang

jamesjwu reviewed (commented):
At a cursory glance, for the specific unit test/feature it's trying to support, this looks pretty good. 

I think it would be good to either add tests or descriptions to the PR to explain the current state of what we can support and what features are planned for the future vs. not going to be supported, i.e.:

- So far, these changes suggest that it only works for inference. (If you try to call .backward(), what happens). I assume we'll add AOTAutograd support next. 

- By the name, I would assume that we plan for fullgraph package to work only for fullgraph=True. But then, maybe I just have a basic question, how is it different from using torch export to begin with? 

If there's a full design doc somewhere for the bigger picture that would be super helpful context.

jamesjwu commented on torch/_fullgraph.py:
```diff
@@ -0,0 +1,52 @@
+
+
+@dataclasses.dataclass
+class _Compilation:
```
Can you annotate these classes with what the API to use them will be, along with some example code?

zhxchen17 committed:
Use AOTI as inductor backend with precompile mode. (#145381)

Summary:

Design doc: https://docs.google.com/document/d/1Z15cBBPjoZ7gH00TSgCdgaYko7a7Br-ERd3_hA-g2IU/edit?usp=sharing

In this diff we are trying to introduce some stateful API to enable a global mode which will force inductor to use AOTI as a backend. Different from PR #141700, we didn't try to populate the package file into caching system, instead we bypass caching to simplify the implementation in the current form.


Similar to PR https://github.com/pytorch/pytorch/pull/141700, I did a quick benchmark to the loading time and it looks like the following:
- Precompile
```
buck run mode/opt scripts/zhxchen17:precompile
```
- Load using cache:
```
time buck run mode/opt scripts/zhxchen17:precompile -- --loader cache
```
Output:
```
real    0m24.593s
user    0m59.342s
sys     0m17.201s
```
- Load using load_fullgraph_package
```
time buck run mode/opt scripts/zhxchen17:precompile -- --loader precompile
```
Output:
```
real    0m10.907s
user    0m9.210s
sys     0m1.173s
```

Test Plan:
buck run mode/opt caffe2/test:test_export -- -r test_fullgraph_package_basic
_function

Differential Revision: D68459341

zhxchen17 commented:
sorry for the poor inline doc for now, I drafted a design doc hopefully to explain a bit: https://docs.google.com/document/d/1Z15cBBPjoZ7gH00TSgCdgaYko7a7Br-ERd3_hA-g2IU/edit?usp=sharing

I will update the inline doc later.

jansel commented on test/export/test_fullgraph_package.py:
```diff
@@ -0,0 +1,49 @@
+        inputs = torch.randn(3, device="cuda"), torch.randn(3, device="cuda")
+        expected = f(*inputs)
+        f = torch.compile(f, fullgraph=True, name=self.id())
+        p, *flags = torch.compiler._enable_fullgraph_package(path=self.path())
```
This API seems a bit clunky, could it be done with a non-mode like API?  I don't understand what the varargs `*flags` part is.

Maybe something like:
```py
f = torch.compiler.cached_compile(f)
...
f.save_compile_artifacts(path)


f.load_compile_artifacts(path)
```

Or even just add `save_compile_artifacts` and `load_compile_artifacts` methids to return value of `torch.compile`

jansel replied:
Happy to chat in more detail if you need more detail on how to implement that.

jansel commented on torch/_fullgraph.py:
```diff
@@ -0,0 +1,52 @@
+    def num_models(self) -> int:
+        return len(self._compilations)
+
+    def num_compilations(self, name: str) -> int:
+        return len(self._compilations[name].paths)
```
Do we need to save multiple compilations in a single file?  This seems kind of weird if we require fullgraph.  The APIs would be simpler (no need for a name) if we had one package per compile.

zhxchen17 replied:
Two cases I can think of:
 1. We need a single package to share the weights from different compilations. E.g. prefill and decoding are compiled differently but they share the same set of weights.
 2. For each python callable, there could be multiple recompilations and we also need a way to map the user provided path to different paths for recompilations if we're want to save one file per compilation.

jansel replied:
Is case 2 supported in this PR?  For case 2 we could auto-generate the name based on the specializations done (we need to have those anyway).

Maybe we could support packages with some sort of container object that could be passed to save:
```
def save_compile_artifacts(self, path_or_package: Union[str, MultiFilePackage]):
   ...
```

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -576,6 +582,39 @@ def compile_fx_inner(
         )
 
 
+_PRECOMPILES: Dict[str, List[Any]] = {}
```
Not thread safe.  I'd think we should be able to eliminate the name and store this on the `torch.compile` object (just need to thread a pointer to that object down into this function).

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -615,6 +654,9 @@ def _compile_fx_inner(
             **graph_kwargs,
         )
 
+    if precompile := _get_precompile(graph_kwargs.get("name"), example_inputs):
+        return precompile
```
This seems like an odd place to check the cache.  By this point we have already run dynamo and AOT Autograd, which incur a lot of compile time -- but then we just throw out the graph we worked so hard to generate.  If we moved this check up to the object returned by `torch.compile` then we could get the compile time down close to zero.

zhxchen17 committed:
Merge branch 'main' into export-D68459341

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -1100,9 +1142,47 @@ def log_graph_runnable() -> str:
+                                        current_callable, 1, graph.device_type
+                                    ).run  # type: ignore[attr-defined]
+                                )
+                            elif graph.device_type == "cpu":
```
Not tested?

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -1100,9 +1142,47 @@ def log_graph_runnable() -> str:
+                                        current_callable, 1
+                                    ).run  # type: ignore[attr-defined]
+                                )
+                            elif graph.device_type == "xpu":
```
Not tested?

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -1100,9 +1142,47 @@ def log_graph_runnable() -> str:
+                            else:
+                                current_callable = compiled_fn
+
+                            if graph.device_type.startswith("cuda"):
```
What about a graph with both CPU and CUDA?

ezyang commented on torch/compiler/__init__.py:
```diff
@@ -461,3 +461,68 @@ def load_cache_artifacts(serialized_artifacts: bytes) -> Optional["CacheInfo"]:
+#       explicitly package precompiled artifacts into a single file.
+#       TODO Eventually we should come up with a context manager style API. To
+#       reduce the complexity of landing changes, we first introduce a set of
+#       stateful interfaces as the future building blocks to begin with.
```
This is going to be intended to be a real API right? Do you think it's too early to put our best foot forward on the API (which includes not having underscores?)
